From 1977d717d2e7263f2a3d14e540f73d6bb8e2d559 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 20 Jan 2022 12:05:57 +0200
Subject: [PATCH] [SW-69581] uart: use device struct for dev specific data

Including device specific headers from common code is not allowed
anymore, and instead the device structures should be used.
Uart device code is refactored to use device structure.

Change-Id: Iae16d26ccc744bd2bc3d9b6397bf640ff2985464
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/app/ext_rom/erom_pcie.c b/app/ext_rom/erom_pcie.c
index 1cbaa59..705c869 100644
--- a/app/ext_rom/erom_pcie.c
+++ b/app/ext_rom/erom_pcie.c
@@ -9,9 +9,16 @@
 #include <pci.h>
 #include <pll.h>
 #include <preboot.h>
+#include <device.h>
+
+static void device_init(void)
+{
+	device_init_uart();
+}
 
 void erom_main_function(void)
 {
+	device_init();
 	/*
 	 * Note: we assume CPU frequency is not changed while CPU is running,
 	 * so the udelay init is done once before we start.
diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index 3665b55..db39f4f 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -265,6 +265,7 @@
 static void device_init(void)
 {
 	device_init_efuse();
+	device_init_uart();
 }
 
 __weak void ppboot_security_init(void) {}
diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index 4bc5b04..3ba446b 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -412,6 +412,7 @@
 static void device_init(void)
 {
 	device_init_efuse();
+	device_init_uart();
 }
 
 __weak void hl_set_reset_cause(uint32_t reset_cause) {}
diff --git a/app/recovery/recovery.c b/app/recovery/recovery.c
index 561e189..661f10e 100644
--- a/app/recovery/recovery.c
+++ b/app/recovery/recovery.c
@@ -7,7 +7,7 @@
 
 #include <stddef.h>
 #include <preboot.h>
-
+#include <device.h>
 #include "cmd_module.h"
 #include "pll.h"
 
@@ -24,9 +24,14 @@
 	},
 };
 
+static void device_init(void)
+{
+	device_init_uart();
+}
+
 void recovery_main(void)
 {
-
+	device_init();
 	udelay_init();
 	hl_uart_init();
 
diff --git a/arch/gaudi/arm/common.c b/arch/gaudi/arm/common.c
index d3d2fa4..ada1a26 100644
--- a/arch/gaudi/arm/common.c
+++ b/arch/gaudi/arm/common.c
@@ -14,6 +14,7 @@
 #include <psoc_gpio1_regs.h>
 #include <psoc_spi_regs.h>
 #include <psoc_gpio3_regs.h>
+#include <psoc_uart_0_regs.h>
 #include <pcie_aux_regs.h>
 #include <pcie_core_regs.h>
 #include <pcie_wrap_regs.h>
@@ -651,3 +652,11 @@
 	efuse.addr_range = EFUSE_READ_RANGE_ALL;
 	efuse.words_per_bank = FUSE_WORDS_PER_BANK;
 }
+
+void device_init_uart(void)
+{
+	uart.uart_data_reg = mmPSOC_UART_0_DATA;
+	uart.uart_state_reg = mmPSOC_UART_0_STATE;
+	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
+	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
+}
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 79fd2cf..d89862c 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -45,6 +45,7 @@
 #include <psoc_arc0_cfg_masks.h>
 #include <psoc_arc1_cfg_regs.h>
 #include <psoc_arc1_cfg_masks.h>
+#include <psoc_uart_0_regs.h>
 #include <pcie_elbi_rr_mstr_if_rr_shrd_lbw_regs.h>
 
 #include "hl.h"
@@ -1220,3 +1221,11 @@
 	efuse.addr_range = EFUSE_READ_RANGE_ALL;
 	efuse.words_per_bank = FUSE_WORDS_PER_BANK;
 }
+
+void device_init_uart(void)
+{
+	uart.uart_data_reg = mmPSOC_UART_0_DATA;
+	uart.uart_state_reg = mmPSOC_UART_0_STATE;
+	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
+	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
+}
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index ccce479..6c4003e 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -10,6 +10,7 @@
 #include <d0_pcie_aux_regs.h>
 #include <d0_psoc_reset_conf_regs.h>
 #include <d0_psoc_efuse_regs.h>
+#include <d0_psoc_uart_0_regs.h>
 #include <d0_parc_arc0_cfg_regs.h>
 #include <gaudi3/gaudi3_boot_info.h>
 
@@ -145,6 +146,14 @@
 	efuse.words_per_bank = FUSE_WORDS_PER_BANK;
 }
 
+void device_init_uart(void)
+{
+	uart.uart_data_reg = mmD0_PSOC_UART_0_DATA;
+	uart.uart_state_reg = mmD0_PSOC_UART_0_STATE;
+	uart.uart_ctrl_reg = mmD0_PSOC_UART_0_CTRL;
+	uart.uart_bauddiv_reg = mmD0_PSOC_UART_0_BAUDDIV;
+}
+
 void lkd_dev_specific_setup(void)
 {
 	cpu_boot_dev_sts_set(CPU_BOOT_DEV_STS_FW_HARD_RST_EN);
diff --git a/arch/goya/arm/common.c b/arch/goya/arm/common.c
index a92ee08..2ece7d5 100644
--- a/arch/goya/arm/common.c
+++ b/arch/goya/arm/common.c
@@ -12,6 +12,7 @@
 #include <psoc_efuse_regs.h>
 #include <psoc_gpio3_regs.h>
 #include <psoc_spi_regs.h>
+#include <psoc_uart_0_regs.h>
 #include <pcie_aux_regs.h>
 #include <pcie_core_regs.h>
 #include <cpu_timestamp_regs.h>
@@ -309,3 +310,11 @@
 	efuse.addr_range = EFUSE_READ_RANGE_ALL;
 	efuse.words_per_bank = FUSE_WORDS_PER_BANK;
 }
+
+void device_init_uart(void)
+{
+	uart.uart_data_reg = mmPSOC_UART_0_DATA;
+	uart.uart_state_reg = mmPSOC_UART_0_STATE;
+	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
+	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
+}
diff --git a/arch/greco/common/common.c b/arch/greco/common/common.c
index f7e221d..2fe7992 100644
--- a/arch/greco/common/common.c
+++ b/arch/greco/common/common.c
@@ -21,6 +21,7 @@
 #include <psoc_arc0_cfg_masks.h>
 #include <psoc_arc1_cfg_regs.h>
 #include <psoc_arc1_cfg_masks.h>
+#include <psoc_uart_0_regs.h>
 #include <pcie_core_regs.h>
 #include <pcie_phy_regs.h>
 #include <pcie_dbi_regs.h>
@@ -589,3 +590,11 @@
 	efuse.addr_range = EFUSE_READ_RANGE_ALL;
 	efuse.words_per_bank = FUSE_WORDS_PER_BANK;
 }
+
+void device_init_uart(void)
+{
+	uart.uart_data_reg = mmPSOC_UART_0_DATA;
+	uart.uart_state_reg = mmPSOC_UART_0_STATE;
+	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
+	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
+}
diff --git a/include/device.h b/include/device.h
index 46810c0..9cf3b6c 100644
--- a/include/device.h
+++ b/include/device.h
@@ -13,6 +13,15 @@
 	u32 words_per_bank;
 };
 
+struct device_uart {
+	u32 uart_data_reg;
+	u32 uart_state_reg;
+	u32 uart_ctrl_reg;
+	u32 uart_bauddiv_reg;
+};
+
 extern struct device_efuse efuse;
+extern struct device_uart uart;
 
 void device_init_efuse(void);
+void device_init_uart(void);
diff --git a/lib/common/uart.c b/lib/common/uart.c
index cf17ec9..5d43f06 100644
--- a/lib/common/uart.c
+++ b/lib/common/uart.c
@@ -6,9 +6,8 @@
  */
 #include <stddef.h>
 
-#include <psoc_uart_0_regs.h>
-
 #include <preboot.h>
+#include <device.h>
 #include <sizes.h>
 #include "uart.h"
 #include "pll.h"
@@ -18,6 +17,8 @@
 #define STR_UINT_SZ			8
 #define BITS_IN_SINGLE_CHAR		4
 #define RING_BUFF_SIZE			SZ_256
+
+struct device_uart uart;
 
 struct ring_buf_t {
 	char buf[RING_BUFF_SIZE];
@@ -60,7 +61,7 @@
 		tx_buf.tail = tail;
 
 		/* transferring data directly to HW */
-		regmap_writel(ch, mmPSOC_UART_0_DATA);
+		regmap_writel(ch, uart.uart_data_reg);
 		preboot_putc_virt_uart(ch);
 	}
 }
@@ -68,7 +69,7 @@
 void preboot_uart_rbuf_flush(void)
 {
 	while (tx_buf.head != tx_buf.tail) {
-		while (regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_TXFULL)
+		while (regmap_readl(uart.uart_state_reg) & UART_SR_TXFULL)
 			;
 		preboot_uart_rbuf_pop();
 	}
@@ -98,14 +99,14 @@
 
 void preboot_uart_init(unsigned int val)
 {
-	regmap_writel(0x0, mmPSOC_UART_0_CTRL);
-	regmap_writel(val, mmPSOC_UART_0_BAUDDIV);
-	regmap_writel(0x3, mmPSOC_UART_0_CTRL);
+	regmap_writel(0x0, uart.uart_ctrl_reg);
+	regmap_writel(val, uart.uart_bauddiv_reg);
+	regmap_writel(0x3, uart.uart_ctrl_reg);
 }
 
 int preboot_uart_tstc(void)
 {
-	if (!(regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_RXFULL))
+	if (!(regmap_readl(uart.uart_state_reg) & UART_SR_RXFULL))
 		return 0;
 
 	return 1;
@@ -116,12 +117,12 @@
 	if (!preboot_uart_tstc())
 		return -1;
 
-	return regmap_readl(mmPSOC_UART_0_DATA);
+	return regmap_readl(uart.uart_data_reg);
 }
 
 bool preboot_uart_tx_full(void)
 {
-	return !!(regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_TXFULL);
+	return !!(regmap_readl(uart.uart_state_reg) & UART_SR_TXFULL);
 }
 
 /* timeout is 5 chars duration in usec*/
@@ -145,9 +146,9 @@
 		return;
 	}
 
-	while (regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_TXFULL)
+	while (regmap_readl(uart.uart_state_reg) & UART_SR_TXFULL)
 		;
-	regmap_writel(ch, mmPSOC_UART_0_DATA);
+	regmap_writel(ch, uart.uart_data_reg);
 	preboot_putc_virt_uart(ch);
 }
 
